KAS-2491: User impersonation

sergiofenoll opened the pull request:
Adds the ability to impersonate users to the app.

- The dropdown in the top right corner has an extra option to select which user (profile) to impersonate
- When impersonating a user the data received is the same as what that user would see
- The user's role is also inherited, so any UI elements that are conditionally rendered based on user role also update accordingly
- Both in the frontend and backend, impersonation is only available for admin users. In the frontend that's done by hiding the impersonation selection for non-admins
- Impersonating a user refreshes the current page you're on so you can immediately see what said user would see on that route. This also means that certain features like default homepage require an extra step to simulate (e.g. OVRB profiles navigate to `/publicaties` by default after log in, to see this with impersonation you would have to impersonate OVRB and then navigate to the root of the app manually)
- Impersonation exists behind the `ENABLE_IMPERSONATION` feature flag. It's enabled by default in development mode as well as on ACC. It should remain toggled off on PROD, as this feature is not desired in that environment.

Related PRs:
- [x] https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/343
- [x] https://github.com/kanselarij-vlaanderen/impersonation-service/pull/1

Jenkins runs:
- http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/333/

sergiofenoll committed:
feat: initial feature commit adding impersonation support

This is very WIP!

sergiofenoll committed:
feat: impersonatee as extra data, session:account is static

sergiofenoll committed:
feat: impersonation endpoints and only get impersonator from service

sergiofenoll committed:
feat: add feature flag for impersonation

sergiofenoll committed:
Merge branch 'development' into feature/KAS-2491-user-impersonation

sergiofenoll committed:
fix: form the body correctly for impersonation

sergiofenoll committed:
ci: checkout matching backend branch when building PR pipeline

If the frontend branch being built by Jenkins exists in the backend,
check it out. Otherwise, check out the branch that we're merging
to (e.g. acceptance). If that also doesn't exist for some reason, just
check out development.

sergiofenoll committed:
feat: impersonation now only switches roles, not users

erikap commented on config/environment.js:
```diff
@@ -40,6 +40,7 @@ module.exports = function (environment) {
     APP: {
       // eslint-disable-next-line quotes
       ENABLE_SIGNATURES: '{{ENABLE_SIGNATURES}}',
+      ENABLE_IMPERSONATION: '{{ENABLE_IMPERSONATION}}',
```
In `Dockerfile` best nog een default zetten voor `EMBER_ENABLE_IMPERSONATION` zodat feature disabled is in builded images by default  (zie ook `EMBER_ENABLE_SIGNATURES` in `Dockerfile`)

erikap commented on app/services/session.js:
```diff
@@ -25,4 +26,9 @@ export default class ExtendedSessionService extends SessionService {
   }
+
+  async invalidate() {
+    await this.impersonation.stopImpersonation();
```
Can't this be added to `currentSession.clear()` which is called in `handleInvalidation()` in this service? As such all impersonation-related data loading/clearing is managed by the `current-session` service.

sergiofenoll replied:
My understanding of the auth addon we use is that `handleInvalidation` only gets called if the underlying session is no longer valid (maybe it expires while the user is logged in). If the user manually logs out, we never go through `handleInvalidation`. I will double check though, because that was last week and perhaps I'm just mistaken.

erikap commented on app/services/impersonation.js:
```diff
@@ -0,0 +1,82 @@
+    }
+  }
+
+  clear() {
```
What's the difference between `clear()` and `stopImpersonation()`?  Shouldn't the logic of both always be executed to keep frontend and backend impersonation state in sync?

erikap commented on app/services/current-session.js:
```diff
@@ -38,21 +44,34 @@ export default class CurrentSessionService extends Service {
+
+  get isReallyAdmin() {
+    return this.ownUserGroup.name == 'ADMIN';
+  }
```
If not used in this PR, I wouldn't introduce this one since we want to move away from this kind of methods and consistently use permissions instead.

erikap commented on app/services/current-session.js:
```diff
@@ -38,21 +44,34 @@ export default class CurrentSessionService extends Service {
     return this.userGroup?.permissions.includes(permission);
   }
 
+  reallyMay(permission) {
+    return this.ownUserGroup?.permissions.includes(permission);
+  }
```
Isn't it more elegant to add an extra option to `may(permission)` which is `false` by default to validate on the impersonator role/group instead? Something like `may(permission, checkImpersonator = false)`

erikap commented on app/services/current-session.js:
```diff
@@ -38,21 +44,34 @@ export default class CurrentSessionService extends Service {
   }
 
   get userGroup() {
+    return this.impersonatedRole && findGroupByRole(this.impersonatedRole.uri);
```
Shouldn't we fallback to `this.role` here in case the user is currently not impersonated?  Or am I missing something?

erikap commented on app/services/current-session.js:
```diff
@@ -7,11 +7,16 @@ import { findGroupByRole } from 'frontend-kaleidos/config/permissions';
-  @tracked role;
   @tracked organization;
   @tracked membership;
+  @tracked ownRole;
```
Doesn't seem to be used?  Shouldn't this still be `@tracked role`?

erikap commented on app/services/current-session.js:
```diff
@@ -7,11 +7,16 @@ import { findGroupByRole } from 'frontend-kaleidos/config/permissions';
+  @tracked ownRole;
+
+  get impersonatedRole() {
+    return this.impersonation.role;
```
Is this 'alias' necessary?  I would like to avoid exposing properties of other services through this service.

erikap commented on app/helpers/user-really-may.js:
```diff
@@ -0,0 +1,10 @@
+import Helper from '@ember/component/helper';
+import { inject as service } from '@ember/service';
+
+export default class UserReallyMay extends Helper {
```
Provide a named argument `checkImpersonator` on the `user-may` helper instead of introducing a separate helper?

erikap commented on app/components/m-header.hbs:
```diff
@@ -58,6 +58,16 @@
     </Auk::Toolbar::Group>
     <Auk::Toolbar::Group @position="right" class="auk-u-mr-3">
+      {{#if
+           (and this.currentSession.impersonatedRole
+           (not-eq this.currentSession.role this.currentSession.impersonatedRole))
```
Provide a `get isImpersonated()` on the `currentSession` service containing this logic to 'hide' service internals

erikap commented on app/components/impersonate-user-modal.js:
```diff
@@ -0,0 +1,30 @@
+import { inject as service } from '@ember/service';
+import { task } from 'ember-concurrency';
+
+export default class ImpersonateUserModalComponent extends Component {
```
Is this component used?

erikap commented on app/components/m-header.hbs:
```diff
@@ -58,6 +58,16 @@
+           (and this.currentSession.impersonatedRole
+           (not-eq this.currentSession.role this.currentSession.impersonatedRole))
+      }}
+        <Auk::Toolbar::Item class="auk-u-mr-3">
```
Can the user stop impersonation without the need to logout?

sergiofenoll replied:
No, though in essence choosing Admin again serves the same purpose. To make it more complete I'll add a proper button though

erikap commented on app/services/impersonation.js:
```diff
@@ -0,0 +1,82 @@
+            relationships: {
+              role: {
+                data: {
+                  type: 'accounts',
```
Must be `roles`?

erikap commented on app/services/impersonation.js:
```diff
@@ -0,0 +1,82 @@
+                data: {
+                  type: 'accounts',
+                  id: role.id,
+                  attributes: {
+                    uri: role.uri,
+                  }
```
Not in line with JSON:API. Is this URI a requirement?

sergiofenoll replied:
Just added it to save on a query in the backend. Probably better not to, then we can check if the passed ID actually exists and if it doesn't the error is a bit more specific.

Aside: can you not pass data about relationships inline in JSON:API? Does relationship data have to go in the `includes` array?

erikap replied:
Afaik, `includes` can only be used in responses, not in requests. You can only pass `type` and `id` for relationships inline in requests.

sergiofenoll committed:
chore: add default value for impersonation flag env var in Dockerfile

sergiofenoll committed:
chore: delete totally unused component

sergiofenoll committed:
chore: cleanup currentSession service

sergiofenoll committed:
fix: stop impersonation in invalidate and handleInvalidation hooks

sergiofenoll committed:
chore: abstract isImpersonated logic behind getter

sergiofenoll committed:
feat: update calls to impersonation service

sergiofenoll committed:
chore: rename isImpersonated to isImpersonator

isImpersonated can sound like "you are being impersonated by someone
else" instead of "you are currently impersonating someone".
isImpersonator shouldn't be ambiguous.

sergiofenoll committed:
feat: add button to stop impersonating roles

sergiofenoll committed:
chore: make checkImpersonator a named argument of user-may helper

erikap committed:
Fix type in JSON:API request for impersonations

erikap committed:
Merge branch 'development' into feature/KAS-2491-user-impersonation

erikap commented:
Jenkins http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/334

erikap committed:
Merge branch 'development' into feature/KAS-2491-user-impersonation

erikap commented:
Failing tests not related to this PR. Going to merge this one.
